File: csharp/20. Valid Parentheses/Solution.cs
public record struct Solution
{
    public bool IsValid(string s)
    {
        Stack<char> stack = new Stack<char>();
        Dictionary<char, char> dictionary = new Dictionary<char, char>(){
            {'{', '}'},
            {'[', ']'},
            {'(', ')'}
        };

        foreach (char character in s)
        {
            if (stack.Count != 0 && dictionary[stack.Peek()] == character)
                stack.Pop();
            else if (!dictionary.ContainsKey(character))
                return false;
            else
                stack.Push(character);
        }

        return stack.Count == 0;
    }
}

File: cpp/20. Valid Parentheses/solution.cpp
#include <stack>
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <tuple>
#include <cassert>

using namespace std;

class Solution
{
public:
    bool isValid(string s)
    {
        stack<char> stack;
        unordered_map<char, char> map{
            {'[', ']'},
            {'{', '}'},
            {'(', ')'},
        };

        for (int i = 0; i < s.size(); ++i)
            if (!stack.empty() && map[stack.top()] == s[i])
                stack.pop();
            else if (!map[s[i]])
                return false;
            else
                stack.push(s[i]);

        return stack.empty();
    }

    void runTests()
    {
        vector<tuple<string, bool>> tests{
            {"[[]]{}(())", true},
            {"[)", false},
            {"[]{{{}}}", true},
            {")()(", false},
            {"{[]}", true}};

        for (int i = 0; i < tests.size(); ++i)
            assert(isValid(get<0>(tests[i])) == get<1>(tests[i]));
    }
};

int main(int argc, char const *argv[])
{
    Solution solution;
    solution.runTests();
    return 0;
}
